### PR TITLE
Reset pulse length, fix ST-Link v2 clone nRST pin

### PR DESCRIPTION
**Pull request by silbe**
## Detailed description

This patch series fixes two issues I had using the BMP firmware on an ST-Link v2 clone with a target that requires using SRST:
- The nRST pulse length is too short and didn't trigger a target reset at all in my case.

  This is addressed by adding an optional parameter to `cmd_reset` specifying the pulse length in milliseconds, defaulting to the previous behaviour of no delay (0ms).
- Unlike the original ST-Link v2 hardware which has nRST connected to PB0 or PB1, the "clones" use PB6.

  Despite having completely different PCB layouts and ICs, all three different types of "clones" I examined don't have nRST hooked up to PB0 or PB1 but rather to PB6. Some have PB6 and PB7 connected together.
  The clones are now assigned their own hardware revision (0x101 to leave plenty of room for more variations of the original hardware). Based on this hardware revision, BMP will now use PB6 as nRST for the clones.

*N.B.*: Building the BMP firmware for ST-Link clones currently requires disabling some of the targets as otherwise it's too large to fit into the 128KiB flash of the clones (build time error). This requires manual changes to `src/Makefile`. For testing this patch series I disabled the `lpc*` targets.

## Your checklist for this pull request

* [x] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md) that has no contact method filled in.
* [x] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [x] It builds for hardware native (`make PROBE_HOST=native`)
* [x] It builds as BMDA (`make PROBE_HOST=hosted`)
* [x] I've tested it to the best of my ability
* [x] My commit messages provide a useful short description of what the commits do

## Closing issues
None.

### CODE REVIEW

**Review by dragonmux — APPROVED** [review]
Unless someone with a ST-Link v2 or one of the clones can chime in saying why we should not merge this as-stands, this looks good to us and we're happy to merge it.

We'll give it a day or two till we pull the trigger on the merge just to let any of the stlink platform users chime in if this somehow breaks things

**Comment by silbe** [comment]
> Unless someone with a ST-Link v2 or one of the clones can chime in saying why we should not merge this as-stands, this looks good to us and we're happy to merge it.
> 
> We'll give it a day or two till we pull the trigger on the merge just to let any of the stlink platform users chime in if this somehow breaks things

Great, thanks! Being careful about not breaking stuff for other users is quite appreciated. :+1: 

> Edit: please fix the whitespace error introduced in the patch to command.c then we'll re-approve this

Done. Also finally flipped the default for C code to use TABs as it's more common across the code bases I'm working with at the moment, plus I installed editorconfig so projects with a corresponding config should use proper settings.

**Review by dragonmux — DISMISSED** [review]

**Comment by silbe** [comment]
Accidentally pushed my local hack for disabling the lpc targets as well. Sorry for that, fixed now.

Should probably add a pre-push hook that recognises these commits but that's not straightforward as there can be only one hook dir, so per-repository hooks would break or need to be chain-loaded from the global hooks directory...

**Comment by dragonmux** [comment]
Checking for unintended changes to the Makefiles is a component of normal review for us anyway as a hook can't tell the difference between unintended changes by way of removals vs quite intentional changes such as a file being renamed, removed on purpose, etc. Easier to do a review and ask questions when there's any uncertainty.

**Review by sidprice — COMMENTED** [review]
It is not clear that the second parameter of the "mon/reset" command is optional. Perhaps change the text to: target: (pulse_len_ms [, default 0])

**Comment by silbe** [comment]
> It is not clear that the second parameter of the "mon/reset" command is optional. Perhaps change the text to: target: (pulse_len_ms [, default 0])

I'm not opposed to change it in general, but I don't think that variant is an improvement. It looks like the default value of the parameter is optional rather than the parameter itself. :wink: 

From what I can tell the parentheses (`(param)`) are already supposed to indicate that the parameter is optional. Most CLI help texts use brackets (`[param]`) for optional parameters so it's not immediately obvious (at least it wasn't to me). The difference between the new `cmd_reset` description and most other commands is that the default value is documented as well. The only other command with documented default value is `halt_timeout` and that one is quite confusing IMO: It looks like leaving the parameter away would set a value of 2000ms while in reality it will print the current value. So I'm not sure how to improve on the new `cmd_reset` description.

I thinks the reason most of the command descriptions are slightly confusing to me is that they are a mix of syntax and semantics.

**Comment by silbe** [comment]
> Checking for unintended changes to the Makefiles is a component of normal review for us anyway as a hook can't tell the difference between unintended changes by way of removals vs quite intentional changes such as a file being renamed, removed on purpose, etc. Easier to do a review and ask questions when there's any uncertainty.

:+1: 

The hook would be a local thing for me; not for BMP. I already labelled the commit as a local hack that's not supposed to go upstream, I just forgot to check before pushing. :blush:

**Comment by dragonmux** [comment]
> From what I can tell the parentheses ((param)) are already supposed to indicate that the parameter is optional. Most CLI help texts use brackets ([param]) for optional parameters so it's not immediately obvious (at least it wasn't to me). The difference between the new cmd_reset description and most other commands is that the default value is documented as well. The only other command with documented default value is halt_timeout and that one is quite confusing IMO: It looks like leaving the parameter away would set a value of 2000ms while in reality it will print the current value. So I'm not sure how to improve on the new cmd_reset description.

In Black Magic Debug, we use `[]` exclusively for optional parameters. This is consistent with BNF, Regex, etc too. It would be consistent to use this markup for the help text, especially as we need to keep the text as short as we reasonably can to conserve Flash space.

**Comment by silbe** [comment]
> In Black Magic Debug, we use `[]` exclusively for optional parameters. This is consistent with BNF, Regex, etc too. It would be consistent to use this markup for the help text, especially as we need to keep the text as short as we reasonably can to conserve Flash space.

If there's consensus on it I'm more than happy to provide a follow-up MR to switch all command help texts to brackets for optional parameters.

**Comment by dragonmux** [comment]
That works for us :)

**Review by dragonmux — APPROVED** [review]

**Comment by dragonmux** [comment]
We'll merge this once the builds complete now we know you've seen the response about the follow-up PR

**Comment by silbe** [comment]
Rebased on `main`; no code changes. Follow-up change to unify help syntax is in #1460.

**Comment by silbe** [comment]
Thanks for the thorough review! Much appreciated.

**Pull request merged**
